rust_snuba/rust_arroyo/src/backends/kafka/mod.rs

```diff
@@ -165,107 +165,113 @@ impl<C: AssignmentCallbacks + Send + Sync> ClientContext for CustomContext<C> {
 }
 
 impl<C: AssignmentCallbacks> ConsumerContext for CustomContext<C> {
-    fn pre_rebalance(&self, base_consumer: &BaseConsumer<Self>, rebalance: &Rebalance) {
-        if let Rebalance::Revoke(list) = rebalance {
-            let mut partitions: Vec<Partition> = Vec::new();
-            for partition in list.elements().iter() {
-                let topic = Topic::new(partition.topic());
-                let index = partition.partition() as u16;
-                partitions.push(Partition::new(topic, index));
-            }
-
-            let mut offsets = self.consumer_offsets.lock().unwrap();
-            for partition in partitions.iter() {
-                offsets.remove(partition);
-            }
-
-            let committer = OffsetCommitter {
-                consumer: base_consumer,
-            };
-
-            self.callbacks.on_revoke(committer, partitions);
-        }
-    }
-
-    fn post_rebalance(&self, base_consumer: &BaseConsumer<Self>, rebalance: &Rebalance) {
-        if let Rebalance::Assign(list) = rebalance {
-            let committed_offsets = base_consumer
-                .committed_offsets((*list).clone(), None)
-                .unwrap();
-
-            let mut offset_map: HashMap<Partition, u64> = HashMap::new();
+    // handle entire rebalancing flow ourselves, so that we can call rdkafka.assign with a
+    // customized list of offsets. if we use pre_rebalance and post_rebalance callbacks from
+    // rust-rdkafka, consumer.assign will be called *for us*, leaving us with this flow on
+    // partition assignment:
+    //
+    // 1. rdkafka.assign done by rdkafka
+    // 2. post_rebalance called
+    // 3. post_rebalance modifies the assignment (if e.g. strict_offset_reset=true and
+    //    auto_offset_reset=latest)
+    // 4. rdkafka.assign is called *again*
+    //
+    // in comparison, confluent-kafka-python will execute on_assign, and only call rdkafka.assign
+    // if the callback did not already explicitly call assign.
+    //
+    // if we call rdkafka.assign multiple times, we have seen random AutoOffsetReset errors popping
+    // up in poll(), since we (briefly) assigned invalid offsets to the consumer
+    fn rebalance(
+        &self,
+        base_consumer: &BaseConsumer<Self>,
+        err: RDKafkaRespErr,
+        tpl: &mut TopicPartitionList,
+    ) {
+        match err {
+            RDKafkaRespErr::RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS => {
```

Comment:
Does this correspond to pre-rebalance or post-rebalance?

Reply:
our pre-rebalance matches for `revoke` directly, so this corresponds to pre-rebalance, and vice-versa down below.

Reply:
we are implementing the function that normally calls both of those callbacks. this function is the entire rebalancing, both pre and post. whatever runs before `consumer.assign` in this function is "pre-rebalance", whatever runs after is "post-rebalance"